### PR TITLE
webtransport: Add helpers to transform webtransport multiaddrs in AddrsFactory

### LINKED ISSUE #2223: Cannot advertise WebTransport with externally known addresses (external static IP + port forwarding, dns, etc.)
When running a go-libp2p host behind a NAT with port forwarding enabled or another environment where the user might know more about the addresses to advertise than go-libp2p has been able to discover (e.g. DNS addresses) users pass in their addresses via the `AddrsFactory` option https://github.com/libp2p/go-libp2p/blob/8c466ab7f8d0e1eb12ea8a23933e77d25190a81a/options.go#L260

As I understand it the `AddrsFactory` takes in the list of addresses that go-libp2p thinks it should be advertising and then allows the user to modify that list (e.g. override/remove elements, add new ones, etc.). The multiaddrs processed and returned by this function are ones that do not have the `/p2p/<peerID>` suffix.

Unfortunately, this behavior doesn't work for WebTransport since naively adding an additional `/ip4/1.2.3.4/udp/1234/quic-v1/webtransport` will be broken due to the lack of WebTransport certificates that are not added internally by go-libp2p. This is not something the user can reasonably deal with by passing the certificates themselves without understanding (and replicating) the internal go-libp2p mechanism for managing WebTransport certificates (having deterministic certificates generated with timestamps based on dividing time into fixed intervals) or finding some awkward workaround (e.g. having a loopback webtransport listener and then reading the certificates go-libp2p has emitted for the loopback listener and feeding it back into the `AddrsFactory`).

### PR DESCRIPTION
**Pull request by MarcoPolo**
I think the AddrsFactory is a bit hard to use. This PR doesn't address that.

This PR adds some helper functions to allow users to do the correct thing inside AddrsFactory. Fixes #2223 

Users can add their own webtransport mappings by copying certhashes from an existing multiaddr. For example,

```go
AddrsFactory(func(multiaddrs []ma.Multiaddr) []ma.Multiaddr {
	outAddrs := make([]ma.Multiaddr, 0, len(multiaddrs)+1)
	for _, addr := range multiaddrs {
		// Keep the original addrs.
		outAddrs = append(outAddrs, addr)
		if webtransport.IsWebtransportMultiaddrWithCerthash(addr) {
			// Add our custom addr with certhashes from an existing addr.
			outAddrs = append(outAddrs, webtransport.CopyCerthashes(addr, customAddr))
		}
	}
	return outAddrs
})
```

This also allows users to do this only for specifics addresses in case they are listening on multiple addresses.

### CODE REVIEW

File: p2p/transport/webtransport/multiaddr.go
Comment (marten-seemann) [inline p2p/transport/webtransport/multiaddr.go:86]: One `iota` should be enough. I don't know how multiple of them behave, does it reset to 0?

File: p2p/transport/webtransport/multiaddr.go
Comment (marten-seemann) [inline p2p/transport/webtransport/multiaddr.go:100]: Why not use an address matcher? We shouldn't be rewriting basic multiaddr functions here.

File: p2p/transport/webtransport/multiaddr.go
Comment (marten-seemann) [inline p2p/transport/webtransport/multiaddr.go:121]: Pre-allocate 2, as this is what we'd usually have.

File: p2p/host/basic/basic_host.go
Comment (marten-seemann) [inline p2p/host/basic/basic_host.go:763]: What's this?

**Review by marten-seemann — COMMENTED** [review]
It's unfortunate that we need this. It's a lot of additional API for something that "should just work". `AddrsFactory` really is a terrbile API, and we should work on cleaning that up soon.

File: p2p/transport/webtransport/multiaddr.go
Comment (MarcoPolo) [reply p2p/transport/webtransport/multiaddr.go:86]: I was following the [docs](https://go.dev/ref/spec#Iota). It's the same thing.

File: p2p/transport/webtransport/multiaddr.go
Comment (MarcoPolo) [reply p2p/transport/webtransport/multiaddr.go:100]: Address matcher doesn't do exactly what I want which is:

1. I don't care if it's ipv4/dns4/ipv6/dns6
2. I don't care if it has certhashes or not, but if it does tell me how many.

I know about address matcher, it just didn't do what I needed. See that hack that I removed in `CanDial`. This is pretty straightforward, but it would be good to expand the address matcher to handle these use cases. Doing that in this PR is out of scope though.

File: p2p/host/basic/basic_host.go
Comment (MarcoPolo) [reply p2p/host/basic/basic_host.go:763]: a typo, thanks.

File: libp2p_test.go
Comment (aschmahmann) [inline libp2p_test.go:324]: What's the reasoning behind surfacing this requirement to any libp2p applications who require something like webtransport + port forwarding rather than just handling it internally (e.g. applying this automatically in the host constructor in config.go)?

Comment (MarcoPolo) [reply libp2p_test.go:324]: I don't think we could do the right thing in 100% of cases. For example, what if the server is listening on two webtransport addresses, which one are you forwarding to (the cerhashes need to line up)? This is something the application using go-libp2p has insight, but not go-libp2p itself. Another example, maybe you also want to return both multiaddrs instead of replacing it.

At a high level though, I think this option `AddrsFactory` is hard to use, and I'd like a better api that covers 90% of use cases in a simpler way.

Comment (aschmahmann) [reply libp2p_test.go:324]: > the cerhashes need to line up

Aren't the certhashes always exactly the same on every webtransport interface you're listening on, so it doesn't matter what interface's certhashes you choose they should all be the same.

https://github.com/libp2p/go-libp2p/blob/8c466ab7f8d0e1eb12ea8a23933e77d25190a81a/p2p/transport/webtransport/cert_manager.go#L63 only takes the private key and time and there's one private key per host and one time globally 😄.

An alternative approach to copy certhashes here could've been to just recalculate the deterministic certhashes and use them, but those functions are unexported in the `webtransport` package (as they probably should be since they're definitely internal implementation details). The way you've implemented this (and me in #2224) happens to be slightly less abstraction breaking.

> At a high level though, I think this option AddrsFactory is hard to use, and I'd like a better api that covers 90% of use cases in a simpler way.

Better APIs sound😄. However, I don't want to rely on future refactors to fix existing issues especially if they could take a while to materialize.

Comment (MarcoPolo) [reply libp2p_test.go:324]: > Aren't the certhashes always exactly the same on every webtransport interface you're listening on, so it doesn't matter what interface's certhashes you choose they should all be the same.

Good point. But this may not always be true. Maybe this is an argument for trying to do this ourselves?

You could have an address that doesn't have certhashes (I don't think we support this right now, but we want to) because it has a valid TLS cert. In that case you would want to forward something without a certhash. Automatically appending a certhash would be wrong.

#2224 was helpful, I took the tests from it. But I didn't like the change where config.go now has special knowledge about WebTransport addresses. I want to keep AddrsFactory as simple as possible.

> However, I don't want to rely on future refactors to fix existing issues especially if they could take a while to materialize.

This change should unblock kubo, no?

Comment (MarcoPolo) [reply libp2p_test.go:324]: Happy to come up with a better change that keeps AddrsFactory simple and makes things better for the user if you have any suggestions!

Comment (aschmahmann) [reply libp2p_test.go:324]: > Good point. But this may not always be true. Maybe this is an argument for trying to do this ourselves?

Yeah, my hope is to ask fewer people to worry about the complexity here assuming it's not much worse in the central place we're dealing with it (e.g. in go-libp2p).

> This change should unblock kubo, no?

Yeah, it will. Technically no changes were needed to unblock kubo as none of the changes made here access private fields/functions, these functions do make things easier though 😄. However, needing to add those extra lines of code seems like to surprise people.

If we went the currently proposed route I'd probably try to save other people, at least in the IPFS ecosystem, the hassle and also add it in a boxo helper so other IPFS implementations don't need to worry too much about it (e.g. https://github.com/ipfs/boxo/pull/230). Maybe that'll help enough 🤷.

> You could have an address that doesn't have certhashes (I don't think we support this right now, but we want to) because it has a valid TLS cert.

Yeah that's a good point. We could certainly do this and that would complicate things, although IIUC it would only cause a problem in a case where:
1. You have an externally provided TLS cert for that address
2. The TLS cert lives outside of go-libp2p (where the constructor couldn't access it) and that part of the QUIC connection establishment happens outside of go-libp2p

Maybe this a reasonable thing. I have no idea how common such a QUIC proxy setup is in general or would be for go-libp2p users. I don't know if people would complain too much if they had to pass extra options to deal with externally established connections.

My preference is to try and eat the complexity so the users don't have to. Given the cases not handled by automatic fixing in the constructor seem to not exist quite yet it seems like we can protect go-libp2p users from thinking about this for a little while longer which is 👍.

---

> Happy to come up with a better change that keeps AddrsFactory simple and makes things better for the user if you have any suggestions!

Unfortunately, nothing better than what I've suggested is coming to mind at the moment. I can see the advantages to both views here, although I have my preference 😄. I'll take whatever the maintainers decide and incorporate with it in the downstreams I help maintain though.

Comment (MarcoPolo) [reply libp2p_test.go:324]: > 2. The TLS cert lives outside of go-libp2p (where the constructor couldn't access it) and that part of the QUIC connection establishment happens outside of go-libp2p

No, we could pass a static tls config to the constructor and that would be fine. This isn't like websockets.

Okay. I think eating this complexity/messiness for the user is the nice thing to do. I'll update this PR.

Concretely:
- If the user returns a webtransport multiaddr that doesn't have a certhash, and all our known addrs have a certhash, copy the certhash over.
  - If we only have a webtransport multiaddr without a certhash (client passed a static tls config), don't copy.
  - If we have both, log a warning and don't copy the certhash.

Does that seem right?

Comment (aschmahmann) [reply libp2p_test.go:324]: Yeah that seems right to me. I suspect if we reach a scenario where someone is trying to instantiate two or more webtransport libp2p transports (e.g. 1+ with CA certs and 1 with the autogenerated ones) and they're having trouble specifying to the `AddrsFactory` how to signal "please give me autogenerated certs" that we can deal with it then.

I couldn't find an issue for enabling being a server with CA certs for webtransport (since they're not currently supported https://github.com/libp2p/go-libp2p/blob/8c466ab7f8d0e1eb12ea8a23933e77d25190a81a/p2p/transport/webtransport/transport.go#L316). If not we should probably start one and then flag this edge case there since this is the kind of thing that could regress or surprise people if the feature isn't implemented in the near future (when any of us still remember this issue 😅).

Comment (MarcoPolo) [reply libp2p_test.go]: https://github.com/libp2p/go-libp2p/issues/2230 is the issue.

I've updated the PR. Thanks for your input @aschmahmann ☺️

File: options.go
Comment (MarcoPolo) [reply options.go:268]: The idea is that this is just middleware. It just wraps the function and the rest of the code doesn't care that this changed.

Another reason to not put it in config.go is that function is already a bit unwieldy

Comment (marten-seemann) [reply options.go:268]: > Technically users can also instantiate directly with `config.NewNode()`
> 
> https://github.com/libp2p/go-libp2p/blob/8c466ab7f8d0e1eb12ea8a23933e77d25190a81a/config/config.go#L288
> 
> although I couldn't find any examples of anyone who did.

Oh no, please don't do that, @aschmahmann! `config` will be removed soon-ish.

File: options.go
Comment (marten-seemann) [inline options.go:280]: I usually don't review whitespace, but this looks really messy ;)

File: options.go
Comment (marten-seemann) [inline options.go:279]: We don't support static TLS configs. As this is just a hack until we have a proper fix, we can reduce complexity here by hardcoding that.

File: p2p/transport/webtransport/multiaddr.go
Comment (marten-seemann) [inline p2p/transport/webtransport/multiaddr.go:120]: Would it make more sense to just have a `AddCerthashes(ma.Multiaddr) ma.Multiaddr` that adds current set of certhashes?

Comment (MarcoPolo) [reply p2p/transport/webtransport/multiaddr.go:120]: It would for the other thing, but not for this. It would need a reference to the
transport which would be annoying. This is a pure function that can be used in
the AddrsFactory with no other context.

File: options.go
Comment (MarcoPolo) [reply options.go:279]: I'll warn instead since it's unexpected.

File: options.go
Comment (MarcoPolo) [reply options.go]: reworded.

File: options.go
Comment (MarcoPolo) [reply options.go]: log.error instead. We shouldn't hit that case, but if we do it's good to highlight it.

File: options.go
Comment (marten-seemann) [inline options.go:293]:
```suggestion
								// This shouldn't be possible, since we don't support static TLS certs
```

**Review by marten-seemann — APPROVED** [review]
LGTM, modulo one suggestion for the comment.

File: options.go
Comment (aschmahmann) [reply options.go:268]: Lol no problem will keep not using it 😜. Was just flagging the code path here as exposed/not having receiving this fix. If you just want to leave it alone your call though.

**Pull request closed without merging**
